### PR TITLE
Adds a FreeRTOS-Plus-TCP module.

### PR DESCRIPTION
**Pull request by mikewolfram**
Fixes part of #447.

### CODE REVIEW

**Comment by mikewolfram** [comment]
I also have the require NetworkInterface.c file, but would add that in the PR of the ETH driver.

**Comment by mikewolfram** [comment]
Oh, and I have -Werror=format enabled, which let's the build fail. Not sure if I rather remove it from my build or add a patch to fix the line in FreeRTOS TCP.

File: ext/aws/FreeRTOSIPConfig.h.in [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+#define ipconfigUDP_MAX_SEND_BLOCK_TIME_TICKS ( 5000 / portTICK_PERIOD_MS )
+
+/* If ipconfigUSE_DHCP is 1 then FreeRTOS+TCP will attempt to retrieve an IP
+address, netmask, DNS server address and gateway address from a DHCP server.  If
+ipconfigUSE_DHCP is 0 then FreeRTOS+TCP will use a static IP address.  The
+stack will revert to using the static IP address even when ipconfigUSE_DHCP is
+set to 1 if a valid configuration cannot be obtained from a DHCP server for any
+reason.  The static configuration used is that passed into the stack by the
+FreeRTOS_IPInit() function call. */
+#define ipconfigUSE_DHCP	{{ options[":::dhcp"] | int }}
```
Comment (salkinium) [inline ext/aws/FreeRTOSIPConfig.h.in]:
```suggestion
#define ipconfigUSE_DHCP	{{ options.dhcp | int }}
```

File: ext/aws/FreeRTOSIPConfig.h.in [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+/* The MTU is the maximum number of bytes the payload of a network frame can
+contain.  For normal Ethernet V2 frames the maximum MTU is 1500.  Setting a
+lower value can save RAM, depending on the buffer management scheme used.  If
+ipconfigCAN_FRAGMENT_OUTGOING_PACKETS is 1 then (ipconfigNETWORK_MTU - 28) must
+be divisible by 8. */
+#define ipconfigNETWORK_MTU		1500
+
+/* Set ipconfigUSE_DNS to 1 to include a basic DNS client/resolver.  DNS is used
+through the FreeRTOS_gethostbyname() API function. */
+#define ipconfigUSE_DNS			{{ options[":::dns"] | int }}
```
Comment (salkinium) [inline ext/aws/FreeRTOSIPConfig.h.in]: or:
```suggestion
#define ipconfigUSE_DNS			{{ options["dns"] | int }}
```
(lbuild name resolution is context aware)

**Comment by salkinium** [comment]
> Not sure if I rather remove it from my build or add a patch to fix the line in FreeRTOS TCP.

You can add a patch file to the freertos-partial repo and make update.py patch it. I do [this for modm-devices too](https://github.com/modm-io/modm-devices/blob/develop/tools/generator/raw-data-extractor/extract-stm32.py#L55-L58).

File: ext/aws/FreeRTOSIPConfig.h.in (hunk 1 above)
Comment (mikewolfram) [reply ext/aws/FreeRTOSIPConfig.h.in]: Learning by doing, copied it from FreeRTOSConfig.h.in. :-)

File: ext/aws/FreeRTOSIPConfig.h.in (hunk 2 above)
Comment (mikewolfram) [reply ext/aws/FreeRTOSIPConfig.h.in]: Above you suggest **options.dhcp**, but here **options["dns"]**? Which one do you prefer?

Comment (salkinium) [reply ext/aws/FreeRTOSIPConfig.h.in]: The shorter one, here `options.dhcp` and `options.dns`. However, this only works with keys that form valid (Python) identifiers, so something like `options.name-with-hypen.or.dot` needs to be `options["name-with-hypen.or.dot"]`.

**Comment by mikewolfram** [comment]
Working on it I was thinking to add something like
```cpp
#if defined __has_include
#if __has_include(<FreeRTOSIPConfigLocal.h>)
#include <FreeRTOSIPConfigLocal.h>
#endif
#endif
```
This would allow the user to add additional defines for FreeRTOS TCP. Or maybe better use a collector in the project.xml file? How would that look like and how to get them into FreeRTOSIPConfig.h?

**Comment by mikewolfram** [comment]
What is needed to get this checks pass?

**Comment by rleh** [comment]
> What is needed to get this checks pass?

Log: https://circleci.com/api/v1.1/project/github/modm-io/modm/20914/output/103/0?file=true&allocation-id=5f45582fa6592e387c3f81cd-0-build%2F172F83F4

```
Error generating documentation for device stm32f103c8t6: In Module 'modm:freertos:tcp': File or directory not found!
In file '../../root/project/ext/aws/module.lb:37' in function 'build':
        env.copy("freertos/FreeRTOS-Plus-TCP", "freertos_plus_tcp",
```

You can run the test locally: `python3 tools/scripts/docs_modm_io_generator.py -t -c -j4`

**Comment by mikewolfram** [comment]
I know. But since that is the result of another pull request, how do I get it into this one? Update the submodule and commit?

**Comment by rleh** [comment]
Yes, you have to update and commit the submodule [`ext/aws/freertos`](https://github.com/modm-io/modm/tree/develop/ext/aws).

**Comment by mikewolfram** [comment]
> Yes, you have to update and commit the submodule [`ext/aws/freertos`](https://github.com/modm-io/modm/tree/develop/ext/aws).

It shows my private repository master branch and I cannot get it changed. Help!!!

File: ext/aws/FreeRTOSIPConfig.h.in [hunk 3]
```diff
@@ -0,0 +1,293 @@
+/******************************************************************************
+ *
+ * MEMC CONFIDENTIAL
+ *
+ * Copyright (C) MEMC Electronic Materials SpA, 2020
+ *
+ * Unauthorized copying of this file, via any medium, is strictly
+ * prohibited.
+ * Proprietary and confidential.
```
Comment (rleh) [inline ext/aws/FreeRTOSIPConfig.h.in]: Where does this file come from?
The copyright notice does not sound like this file is legally usable in modm.

**Review by rleh — DISMISSED** [review]

File: ext/aws/FreeRTOSIPConfig.h.in (hunk 3 above)
Comment (rleh) [reply ext/aws/FreeRTOSIPConfig.h.in]: [This official "default" config file](https://github.com/FreeRTOS/FreeRTOS/blob/master/FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/include/FreeRTOSIPConfigDefaults.h) is MIT licensed.

Comment (mikewolfram) [reply ext/aws/FreeRTOSIPConfig.h.in]: Autsch, sorry, created a new file which gets automatically that header and then copied the remaining part. Will fix it.

Comment (mikewolfram) [reply ext/aws/FreeRTOSIPConfig.h.in]: Fixed

**Comment by salkinium** [comment]
I've squashed your commits.

> This would allow the user to add additional defines for FreeRTOS TCP. Or maybe better use a collector in the project.xml file? How would that look like and how to get them into FreeRTOSIPConfig.h?

This is great idea, so I added this to both configs. Please check if this works for you, then I'd like to merge this PR.

(lbuild collectors are only useful when additional side-effects from modm are required, for example, the build system flags are reformatted into the specific build systems. For this, it's completely overkill, your solution is way better.)

File: ext/aws/module.lb [hunk 4]
```diff
@@ -11,6 +11,35 @@
...
+        module.add_option(
+	        BooleanOption(
+	            name="dhcp",
+	            description="Use DHCP",
+	            default=False))
+        module.add_option(
+	        BooleanOption(
+	            name="dns",
+	            description="Enable simple DNS client/resolver",
+	            default=False))
```
Comment (salkinium) [inline ext/aws/module.lb]: Are these options required in lbuild, or can be defined via `FreeRTOSIPConfigLocal.h`?

Comment (mikewolfram) [reply ext/aws/module.lb]: I came up with the idea of the local config file after adding the options. Could be removed.

**Comment by mikewolfram** [comment]
Is it ok to assume that the compile has the __has_include() function? What if not?

**Comment by salkinium** [comment]
> Is it ok to assume that the compile has the __has_include() function? What if not?

Yes, it's part of C++17 and both GCC and LLVM have it implemented.

**Comment by salkinium** [comment]
I've removed the options and added some documentation, does this work for you?

**Comment by mikewolfram** [comment]
Now a test is failing, what is wrong?

**Review by salkinium — APPROVED** [review]

**Comment by salkinium** [comment]
> Now a test is failing, what is wrong?

The arm-eabi-gcc download link isn't working in the CI. It's working for me locally, so I'm just going to ignore it for now.

**Comment by salkinium** [comment]
(The macOS CI is not part of the required tests, since it has been proven a bit unreliable.)

**Pull request merged**